src/Features/Core/Portable/EditAndContinue/Remote/RemoteEditAndContinueServiceProxy.cs

```diff
@@ -0,0 +1,462 @@
+            if (client == null)
+            {
+                GetLocalService().EndDebuggingSession(out documentsToReanalyze);
+            }
+            else
+            {
+                documentsToReanalyze = await client.RunRemoteAsync<ImmutableArray<DocumentId>>(
+                    WellKnownServiceHubService.RemoteEditAndContinueService,
+                    nameof(IRemoteEditAndContinueService.EndDebuggingSessionAsync),
+                    solution: null,
+                    Array.Empty<object>(),
+                    callbackTarget: null,
+                    cancellationToken).ConfigureAwait(false);
+            }
+
+            // clear emit/apply diagnostics reported previously:
+            diagnosticUpdateSource.ClearDiagnostics();
+
+            // clear diagnostics reported during run mode:
+            diagnosticService.Reanalyze(_workspace, documentIds: documentsToReanalyze);
+        }
+
+        public async Task<ImmutableArray<Diagnostic>> GetDocumentDiagnosticsAsync(Document document, DocumentActiveStatementSpanProvider activeStatementSpanProvider, CancellationToken cancellationToken)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                return await GetLocalService().GetDocumentDiagnosticsAsync(document, activeStatementSpanProvider, cancellationToken).ConfigureAwait(false);
+            }
+
+            var diagnosticData = await client.RunRemoteAsync<ImmutableArray<DiagnosticData>>(
+                WellKnownServiceHubService.RemoteEditAndContinueService,
+                nameof(IRemoteEditAndContinueService.GetDocumentDiagnosticsAsync),
+                document.Project.Solution,
+                new object[] { document.Id },
+                callbackTarget: new DocumentActiveStatementProviderCallback(activeStatementSpanProvider),
+                cancellationToken).ConfigureAwait(false);
+
+            using var _ = ArrayBuilder<Diagnostic>.GetInstance(out var result);
+            foreach (var data in diagnosticData)
+            {
+                result.Add(await data.ToDiagnosticAsync(document.Project, cancellationToken).ConfigureAwait(false));
+            }
+
+            return result.ToImmutable();
+        }
+
+        public async Task<bool> HasChangesAsync(Solution solution, SolutionActiveStatementSpanProvider activeStatementSpanProvider, string sourceFilePath, CancellationToken cancellationToken)
+        {
+            var client = await RemoteHostClient.TryGetClientAsync(_workspace, cancellationToken).ConfigureAwait(false);
+            if (client == null)
+            {
+                return await GetLocalService().HasChangesAsync(solution, activeStatementSpanProvider, sourceFilePath, cancellationToken).ConfigureAwait(false);
+            }
```

Comment:
seeing this all over the place, i def think i woudl have preferred a single place we get teh client, that we then can use uniformly to determine which direction to send the message to.  it could even be wrapped with an interface so this just forwards to the client the .HasChangesAsync call.  the local version does this, and the remote version does the RunRemoteAsync.

Reply:
Do you mean having something like CreateAsync(Workspace, CT) that retrieves the client and then creates RemoteEditAndContinueServiceProxy with client set in a field? That would remove the calls to GetClient from these methods, but would add extra complexity to the callers of the proxy since most of them create the proxy in the constructor and can't use async there. This approach (of a facade) is conceptually simpler.